microprofile-fault-tolerance/src/test/java/org/jboss/eap/qe/microprofile/fault/tolerance/UndeployDeployTest.java

```diff
@@ -0,0 +1,304 @@
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.wildfly.extras.creaper.core.online.OnlineManagementClient;
+
+/**
+ * Tests deploy/undeploy of MP FT deployments (WAR archives)
+ */
+@RunAsClient
+@RunWith(Arquillian.class)
+public class UndeployDeployTest {
+
+    private static final String FIRST_DEPLOYMENT = "UndeployDeployTest-first-deployment";
+    private static final String SECOND_DEPLOYMENT = "UndeployDeployTestsecond-deployment";
+    private static final String NO_MP_FT_DEPLOYMENT = "no-mp-ft-deployment";
+
+    @ArquillianResource
+    private Deployer deployer;
+
+    @Deployment(name = FIRST_DEPLOYMENT, managed = false)
+    public static Archive<?> createFirstDeployment() {
+        String mpConfig = "hystrix.command.default.execution.timeout.enabled=true";
+
+        return ShrinkWrap.create(WebArchive.class, FIRST_DEPLOYMENT + ".war")
+                .addPackages(true, HelloService.class.getPackage())
+                .addAsManifestResource(EmptyAsset.INSTANCE, "beans.xml")
+                .addAsManifestResource(new StringAsset(mpConfig), "microprofile-config.properties");
+    }
+
+    @Deployment(name = SECOND_DEPLOYMENT, managed = false)
+    public static Archive<?> createSecondDeployment() {
+        String mpConfig = "hystrix.command.default.execution.timeout.enabled=false";
+
+        return ShrinkWrap.create(WebArchive.class, SECOND_DEPLOYMENT + ".war")
+                .addPackages(true, HelloService.class.getPackage())
+                .addAsManifestResource(EmptyAsset.INSTANCE, "beans.xml")
+                .addAsManifestResource(new StringAsset(mpConfig), "microprofile-config.properties");
+    }
+
+    @Deployment(name = NO_MP_FT_DEPLOYMENT, managed = false, testable = false)
+    public static Archive<?> createNonMPFTDeployment() {
+        return ShrinkWrap.create(WebArchive.class, NO_MP_FT_DEPLOYMENT + ".war")
+                .addPackage(org.jboss.eap.qe.microprofile.fault.tolerance.deployments.nofaulttolerance.HelloService.class
+                        .getPackage())
+                .addAsManifestResource(EmptyAsset.INSTANCE, "beans.xml");
+    }
+
+    @BeforeClass
+    public static void setup() throws Exception {
+        MicroProfileFaultToleranceServerConfiguration.enableFaultTolerance();
+    }
+
+    /**
+     * Deploy all deployments so it's possible to get their URLs in other tests. This is limitation of Arquillian
+     * when deployments are deployed manually (in tests) and thus their URL is not known in time when test is started.
+     * The only way how to workaround this issue is to deploy all deployments in first test which allows to inject
+     * deployment URL as params on test methods.
+     */
+    @Test
+    @InSequence(1)
+    public void deployAll() {
```

Comment:
Why are you deploying in advance here? I see each test method wil manage it's deployment in his own. Is it because having to deply manually, then `@OperateOnDeplyment` would fail?

Reply:
Discussed on chat. I've updated comment. 

fixed

Reply:
Right but I stil have one question. If the flow is:
```
1. deploy a, b, c
10. deploy a
...
```
Isn't step #10 (which is using MP FT) causing a warning for already configured Hystrix as in previous tests?

Reply:
No because there is:
```
 @After
    public void undeploy() {
        // in case of failure, clean up deployments
        deployer.undeploy(FIRST_DEPLOYMENT);
        deployer.undeploy(SECOND_DEPLOYMENT);
        deployer.undeploy(NO_MP_FT_DEPLOYMENT);
    }
```
and it causes that Hystrix is shutdown as there is no MP FT deployment using it.

Reply:
Right. Here is what I missed. I think that I was still somehow relating that code to the `@AfterClass` method, thanks for the hint.
That said, if I understand correctly, the `@After` method will be executed after each test method, thus trying to undeploy _all_ of the three deployments each time, but actually finding all of them there only after `InSequence(1)` and - also - will be executed not just `in case of failure` as the comment states, right?

Reply:
Yes, main purpose of after class is to undeploy all deployments if they are present (there is no error if such deployment does not exist = was not deployed) and clean environment for next test no matter what happens.

Reply:
Great, `no matter what happens` and `no error if such deployment does not exist` were important to understand the reasoning behind the implementation you provided for this very case.